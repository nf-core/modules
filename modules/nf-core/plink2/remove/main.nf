process PLINK2_REMOVE {
    tag "$meta.id"
    label 'process_low'
   
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/plink2:2.00a5.10--h4ac6f70_0':
        'biocontainers/plink2:2.00a5.10--h4ac6f70_0' }"

    input:
    tuple val(meta), path(plink_genotype_file), path(plink_variant_file), path(plink_sample_file)
    path(sample_exclude_list)

    output:
    tuple val(meta), path("*.bim"),  emit: remove_bim,  optional: true
    tuple val(meta), path("*.bed"),  emit: remove_bed,  optional: true
    tuple val(meta), path("*.fam"),  emit: remove_fam,  optional: true
    tuple val(meta), path("*.pgen"), emit: remove_pgen, optional: true
    tuple val(meta), path("*.psam"), emit: remove_psam, optional: true
    tuple val(meta), path("*.pvar"), emit: remove_pvar, optional: true
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def mode = plink_genotype_file.extension == "pgen" ? '--pfile' : '--bfile'
    def outtype = plink_genotype_file.extension == "pgen" ? '--make-pgen' : '--make-bed'
    def input = "${plink_genotype_file.getBaseName()}"
    if( "${input}" == "${prefix}" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

    """
    plink2 \\
        $mode $input \\
        $args \\
        --threads $task.cpus \\
        --remove $sample_exclude_list \\
        $outtype \\
        --out $prefix

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        plink2: \$(plink2 --version 2>&1 | sed 's/^PLINK v//; s/ 64.*\$//' )
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def input = "${plink_genotype_file.getBaseName()}"
    def trio = plink_genotype_file.extension == 'pgen' ? "${prefix}.pgen ${prefix}.psam ${prefix}.pvar" : "${prefix}.bed ${prefix}.bim ${prefix}.fam"
    if( "${input}" == "${prefix}" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

    """  
    touch ${trio}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        plink2: \$(plink2 --version 2>&1 | sed 's/^PLINK v//; s/ 64.*\$//' )
    END_VERSIONS
    """
}
