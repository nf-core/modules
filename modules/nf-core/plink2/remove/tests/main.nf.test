nextflow_process {

    name "Test Process PLINK2_REMOVE"
    script "../main.nf"
    process "PLINK2_REMOVE"

    tag "modules"
    tag "modules_nfcore"
    tag "plink2"
    tag "plink2/remove"
    tag "gawk"

    test("human - bed/bim/fam - remove") {

        when {
            process {
                """          
                 input[0] = [
                        [id:"test"],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.bed', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.bim', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.fam', checkIfExists: true)
                ]
                input[1] = Channel.of('per0').collectFile(name:"remove.txt")
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("human - pgen/pvar/psam - remove") {

        when {
            process {
                """          
                 input[0] = [
                        [id:"test"],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.pgen', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.pvar', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.psam', checkIfExists: true)
                ]
                input[1] = Channel.of('per0').collectFile(name:"remove.txt")
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
    
     test("human - bed/bim/fam - remove - stub") {

        options "-stub"

        when {
            process {
                """          
                 input[0] = [
                        [id:"test"],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.bed', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.bim', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.fam', checkIfExists: true)
                ]
                input[1] = Channel.of('per0').collectFile(name:"remove.txt")
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("human - pgen/pvar/psam - remove - stub") {

        options "-stub"

        when {
            process {
                """          
                 input[0] = [
                        [id:"test"],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.pgen', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.pvar', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/popgen/plink_simulated.psam', checkIfExists: true)
                ]
                input[1] = Channel.of('per0').collectFile(name:"remove.txt")
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }   
    
}