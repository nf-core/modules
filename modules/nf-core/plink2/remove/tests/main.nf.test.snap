{
    "human - bed/bim/fam - remove - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bim:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ],
                "remove_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "remove_bim": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bim:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "remove_fam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "remove_pgen": [
                    
                ],
                "remove_psam": [
                    
                ],
                "remove_pvar": [
                    
                ],
                "versions": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-28T14:02:38.833555"
    },
    "human - bed/bim/fam - remove": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bim:md5,41351028a80a8cf5bec0ab2db04af6c9"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,e1f17b70b8f213aa6d48dbde955451b4"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fam:md5,56314a08dc651daea82dd5b04da02d31"
                    ]
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ],
                "remove_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,e1f17b70b8f213aa6d48dbde955451b4"
                    ]
                ],
                "remove_bim": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bim:md5,41351028a80a8cf5bec0ab2db04af6c9"
                    ]
                ],
                "remove_fam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fam:md5,56314a08dc651daea82dd5b04da02d31"
                    ]
                ],
                "remove_pgen": [
                    
                ],
                "remove_psam": [
                    
                ],
                "remove_pvar": [
                    
                ],
                "versions": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-28T14:02:20.262972"
    },
    "human - pgen/pvar/psam - remove": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pgen:md5,d37c6d0aa92f788705e98813907d197a"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.psam:md5,dcecc49a20b9e2b26fe9883789908203"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pvar:md5,723cab01a7556da85d3594c1637ee5d4"
                    ]
                ],
                "6": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ],
                "remove_bed": [
                    
                ],
                "remove_bim": [
                    
                ],
                "remove_fam": [
                    
                ],
                "remove_pgen": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pgen:md5,d37c6d0aa92f788705e98813907d197a"
                    ]
                ],
                "remove_psam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.psam:md5,dcecc49a20b9e2b26fe9883789908203"
                    ]
                ],
                "remove_pvar": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pvar:md5,723cab01a7556da85d3594c1637ee5d4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-28T14:02:29.601086"
    },
    "human - pgen/pvar/psam - remove - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pgen:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.psam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pvar:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ],
                "remove_bed": [
                    
                ],
                "remove_bim": [
                    
                ],
                "remove_fam": [
                    
                ],
                "remove_pgen": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pgen:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "remove_psam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.psam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "remove_pvar": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pvar:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,32ee1177fa4e419aa2f72daae10ff3cf"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-28T14:55:51.6974"
    }
}